strawberryfields/decompositions.py

```diff
@@ -400,44 +414,139 @@ def mach_zehnder(m, n, internal_phase, external_phase, nmax):
-    by the first phaseshifter (before the first beamsplitter) is named
-    ``external_phase``, while we call the phase shift between the beamsplitters
-    ``internal_phase``.
+    return [n, n + 1, phi_i, phi_e, nmax]
 
-    The algorithm applied in this function makes use of the following identity:
 
-    .. code-block:: python
+def nullMZ(n, m, U):
+    r"""Nullifies element n,m of U using mach_zehnder.
 
-        Rgate(alpha) | 1
-        Rgate(beta) | 2
-        Rgate(phi) | 1
-        BSgate(theta, 0) | 1, 2
+    Args:
+        n (int): row index of element to be nullified
+        m (int): column index of element to be nullified
+        U (array): matrix whose m,n element is to be nullified
 
-        equals
+    Returns:
+        list: list containing ``[m, n, internal_phase, external_phase, nmax]`` of the
+            mach_zehnder unitaries needed
+    """
+    (nmax, mmax) = U.shape
 
-        Rgate(phi+alpha-beta) | 1
-        BSgate(pi/4, pi/2) | 1, 2
-        Rgate(2*theta+pi) | 1, 2
-        BSgate(pi/4, pi/2) | 1, 2
-        Rgate(beta-theta+pi) | 1
-        Rgate(beta-theta) | 2
+    if nmax != mmax:
+        raise ValueError("U must be a square matrix")
+
+    if U[n, m] == 0:
+        phi_i = np.pi
+        phi_e = 0
+    elif U[n - 1, m] == 0:
+        phi_i = 0
+        phi_e = 0
+    else:
+        r = U[n - 1, m] / U[n, m]
+        phi_i = 2 * np.arctan(np.abs(r))
+        phi_e = -np.angle(r)
+
+    return [n - 1, n, phi_i, phi_e, nmax]
 
-    The phase-shifts by ``alpha`` and ``beta`` are thus pushed consecutively through
-    all the T unitaries of the interferometer and these unitaries are converted
-    into pairs of symmetric beamsplitters with two phase shifts. The phase
-    shifts at the end of the interferometer are added to the ones from the
-    diagonal unitary at the end of the interferometer obtained from :func:`~.rectangular_phase_end`.
+
+def rectangular_MZ(V, tol=1e-11):
+    r"""Rectangular decomposition of a unitary matrix, with local
+    phase shifts applied between two interferometers.
+
+    Is similar to :func:`rectangular` except that it uses mach_zehner matrices to null elements of V
+    using the :func:`null_MZ` and :func:`null_MZi` instead of :func:`T` matrices and corresponding :func:`nullT`
+    and :func:`nullTi` functions.
+
+    Args:
+        V (array[complex]): unitary matrix of size n_size
+        tol (float): the tolerance used when checking if the matrix is unitary
+
+    Returns:
+        tuple[array]: tuple of the form ``(tilist,np.diag(localV),tlist)``
+            where:
+
+            * ``tilist``: list containing ``[n,m,phi_int,phi_ext,n_size]`` of the ``mach_zehnder_inv`` unitaries needed
+            * ``tlist``: list containing ``[n,m,phi_int,phi_ext,n_size]`` of the ``mach_zehnder`` unitaries needed
+            * ``localV``: Diagonal unitary sitting sandwiched by ``mach_zehnder_inv``'s and the ``mach_zehnder``'s
```

Comment:
```suggestion
            * ``localV``: diagonal unitary sandwiched between the ``mach_zehnder_inv`` unitaries and the ``mach_zehnder`` unitaries
```

A quirk of Sphinx/ReST markup is that you must always have spaces after backticks 🙁